scripts/us_census/acs5yr/subject_tables/common/generateColMap.py

```diff
@@ -0,0 +1,94 @@
+import statvar_dcid_generator
+
+dcid_gen = statvar_dcid_generator.StatVarDcidGenerator()
+
+class generateColMapBase:
+  """module to generate a column map given a list of columns of the dataset and a JSON Spec"""
+  def __init__(self, specDict=None, columnList=None):
+    """module init """
+    self.features = specDict
+    self.columnList = columnList
+    self.columnMap = {}
+    self.makeColMap()
+
+  def makeColMap(self):
+    #for each column generate the definition of their respective statistical variable node
+    for col in self.columnList:
+      if 'ignoreColumns' in self.features and col not in self.features['ignoreColumns']:
+          self.columnMap[col] = self.column_to_statVar(col)
+    #TODO: Enable validating the generate columnMap
+    #self.valid_colMap():
+    return self.columnMap
+
+  def column_to_statVar(self, column):
+    """generates a dictionary statistical variable with all properties specified in the JSON spec for a single column"""
+    base = False
+    statVar = {}
+    partList = column.split('!!')
+
+    for part in partList:
+      #set the base for special cases like median, etc.
+      if not base and 'measurement' in self.features:
+        if part in self.features['measurement']:
+          statVar.update(self.features['measurement'][part])
+          base = True
+
+    #set the default statVar definition
+    if not base and 'measurement' in self.features and '_DEFAULT' in self.features['measurement']:
+      statVar.update(self.features['measurement']['_DEFAULT'])
+
+    #set the populationType attribute
+    statVar['populationType'] = self.get_populationType(partList)
+
+    # associate pvs to statVar
+    if partList[-1] not in self.features['measurement']:
+      if 'pvs' in self.features:
+        #p = property, k = dictionary of partList[-1] to propertyValue
+        for p, k in self.features['pvs'].items():
+          for c, v in k.items():
+            if c == partList[-1]: #last token of column name matches in property
+              statVar[p] = v
+              if 'inferredSpec' in self.features and p in self.features['inferredSpec']:
+                statVar.update(self.features['inferredSpec'][p])
+            ##if the property is somewhere in the column name
+            for part in partList:
+              if c in part:
+                statVar[p] = v
+
+    #TODO: Add units and scalingFactor
+
+    #if the column is Margin of Error, update statType key
+    if 'Margin of Error' in partList:
+      statVar['statType'] = 'marginOfError'
+
+    #generating dcid using the utils/statvar_dcid_generator.py
+    statVar['dcid'] = dcid_gen.get_stat_var_name(statVar, ignore_props=None)
+
+    ##Add Universe PVs based on the populationType of StatVar
+    if 'universePVs' in self.features:
+      for elem in self.features['universePVs']:
+        if statVar['populationType'] == elem['populationType']:
+          for k, v in elem['dependentPVs'].items():
```

Comment:
I thought universePVs are those that will show up in statVar and need to removed while generate SV ID, so those props should be in ignore_props in the call to get_stat_var_name().

Also universePVs might depend on constraints too.  And we should also validate a universePV is actually one by comparing the P and the V.  something like https://source.corp.google.com/piper///depot/google3/datacommons/prophet/flume_generator/stat_var_group_util.cc;l=184-194

Reply:
Good point! I did not check-in the code with `constraintProperties` validation.
I will be checking the validation in the next commit.